File: src/mock/quotations.ts
/**
 * 报价单模拟数据
 * - inq-4（部分已报价）：仅 sup-5 提交
 * - inq-5（报价已完成）：3 家全部提交，含异常（sup-3 报价金额异常高、sup-5 交货周期异常长）
 * - inq-6（已超时）：受邀供应商报价状态为 TIMEOUT
 * - inq-7（已完成）：2 家提交，对应已定标结果
 */
import dayjs from 'dayjs';
import { QuotationStatus, type Quotation, type QuotationItem } from '@/types';

const now = dayjs();
const ts = (offsetDays: number, hour = 10, minute = 0): string =>
  now.add(offsetDays, 'day').hour(hour).minute(minute).second(0).format('YYYY-MM-DD HH:mm:ss');
const dateOnly = (offset: number): string => now.add(offset, 'day').format('YYYY-MM-DD');

let itemSeq = 0;
/** 构造报价明细（unitPrice 视为含税单价，taxIncludedTotal = unitPrice * quantity） */
const buildItem = (
  quotationId: string,
  inquiryItemId: string,
  unitPrice: number,
  quantity: number,
  deliveryDays: number,
  brand: string,
  warrantyMonths = 12,
  paymentTerms = '货到验收后 30 天付款',
): QuotationItem => {
  itemSeq += 1;
  return {
    id: `qitem-${itemSeq}`,
    quotationId,
    inquiryItemId,
    unitPrice,
    taxRate: 0.13,
    taxIncludedTotal: Number((unitPrice * quantity).toFixed(2)),
    moq: 1,
    deliveryDays,
    deliveryDate: dateOnly(deliveryDays),
    brand,
    warrantyMonths,
    paymentTerms,
    validUntil: dateOnly(30),
    attachments: [],
  };
};

/** 供应商名称映射 */
const SUP_NAME: Record<string, string> = {
  'sup-1': '上海恒远工业设备有限公司',
  'sup-2': '苏州联创自动化科技有限公司',
  'sup-3': '宁波华泰五金制造有限公司',
  'sup-4': '深圳智联电子科技有限公司',
  'sup-5': '杭州启明供应链有限公司',
  'sup-6': '广东正达包装材料有限公司',
};

/** 报价单列表 */
export const quotations: Quotation[] = [
  // ===== inq-4 部分已报价：仅 sup-5 提交 =====
  {
    id: 'quo-4-5',
    inquiryId: 'inq-4',
    supplierId: 'sup-5',
    supplierName: SUP_NAME['sup-5'],
    status: QuotationStatus.SUBMITTED,
    submittedAt: ts(-2, 14, 30),
    items: [
      buildItem('quo-4-5', 'item-inq-4-1', 4380, 30, 12, '联想', 36, '货到验收后 30 天付款'),
    ],
    totalAmount: 131400,
    remark: '含预装系统及办公软件，提供上门安装服务。',
    attachments: [],
    createdAt: ts(-4, 9, 0),
    updatedAt: ts(-2, 14, 30),
  },

  // ===== inq-5 报价已完成：3 家全部提交，含异常 =====
  {
    id: 'quo-5-1',
    inquiryId: 'inq-5',
    supplierId: 'sup-1',
    supplierName: SUP_NAME['sup-1'],
    status: QuotationStatus.SUBMITTED,
    submittedAt: ts(-4, 16, 0),
    items: [
      buildItem('quo-5-1', 'item-inq-5-1', 32, 200, 7, 'SKF', 24),
      buildItem('quo-5-1', 'item-inq-5-2', 60, 50, 7, '华泰', 24),
    ],
    totalAmount: 9400,
    remark: '战略供应商价格，含材质证明及检验报告。',
    attachments: [],
    createdAt: ts(-6, 9, 0),
    updatedAt: ts(-4, 16, 0),
  },
  {
    // 异常：报价金额显著高于平均水平（轴承单价 80 元，远高于其他供应商约 33 元）
    id: 'quo-5-3',
    inquiryId: 'inq-5',
    supplierId: 'sup-3',
    supplierName: SUP_NAME['sup-3'],
    status: QuotationStatus.SUBMITTED,
    submittedAt: ts(-2, 10, 0),
    items: [
      buildItem('quo-5-3', 'item-inq-5-1', 80, 200, 12, '替代品牌', 12),
      buildItem('quo-5-3', 'item-inq-5-2', 70, 50, 12, '华泰', 12),
    ],
    totalAmount: 19500,
    remark: '轴承需外调，价格略高；法兰可现货供应。',
    attachments: [],
    createdAt: ts(-5, 9, 0),
    updatedAt: ts(-2, 10, 0),
  },
  {
    // 异常：交货周期显著长于平均水平（40 天，较平均约 20 天多 20 天以上）
    id: 'quo-5-2',
    inquiryId: 'inq-5',
    supplierId: 'sup-5',
    supplierName: SUP_NAME['sup-5'],
    status: QuotationStatus.SUBMITTED,
    submittedAt: ts(-3, 15, 0),
    items: [
      buildItem('quo-5-2', 'item-inq-5-1', 34, 200, 40, '替代品牌', 12),
      buildItem('quo-5-2', 'item-inq-5-2', 63, 50, 40, '华泰', 12),
    ],
    totalAmount: 9950,
    remark: '部分物料需调配，交货周期较长。',
    attachments: [],
    createdAt: ts(-5, 14, 0),
    updatedAt: ts(-3, 15, 0),
  },

  // ===== inq-6 已超时：受邀供应商报价状态为 TIMEOUT =====
  {
    id: 'quo-6-1',
    inquiryId: 'inq-6',
    supplierId: 'sup-6',
    supplierName: SUP_NAME['sup-6'],
    status: QuotationStatus.TIMEOUT,
    submittedAt: undefined,
    items: [],
    totalAmount: 0,
    remark: '未在截止时间前提交报价。',
    attachments: [],
    createdAt: ts(-10, 9, 0),
    updatedAt: ts(-5, 18, 0),
  },
  {
    id: 'quo-6-2',
    inquiryId: 'inq-6',
    supplierId: 'sup-5',
    supplierName: SUP_NAME['sup-5'],
    status: QuotationStatus.TIMEOUT,
    submittedAt: undefined,
    items: [],
    totalAmount: 0,
    remark: '未在截止时间前提交报价。',
    attachments: [],
    createdAt: ts(-10, 9, 0),
    updatedAt: ts(-5, 18, 0),
  },

  // ===== inq-7 已完成：2 家提交，对应定标结果 =====
  {
    id: 'quo-7-2',
    inquiryId: 'inq-7',
    supplierId: 'sup-2',
    supplierName: SUP_NAME['sup-2'],
    status: QuotationStatus.SUBMITTED,
    submittedAt: ts(-15, 16, 0),
    items: [
      buildItem('quo-7-2', 'item-inq-7-1', 880, 15, 10, '华为', 36),
      buildItem('quo-7-2', 'item-inq-7-2', 4100, 8, 12, '西门子', 36),
    ],
    totalAmount: 46000,
    remark: '含编程调试服务，PLC 提供现场技术支持。',
    attachments: [],
    createdAt: ts(-17, 9, 0),
    updatedAt: ts(-15, 16, 0),
  },
  {
    id: 'quo-7-4',
    inquiryId: 'inq-7',
    supplierId: 'sup-4',
    supplierName: SUP_NAME['sup-4'],
    status: QuotationStatus.SUBMITTED,
    submittedAt: ts(-14, 15, 0),
    items: [
      buildItem('quo-7-4', 'item-inq-7-1', 820, 15, 9, '华为', 36),
      buildItem('quo-7-4', 'item-inq-7-2', 4300, 8, 11, '西门子', 36),
    ],
    totalAmount: 46700,
    remark: '交换机价格优势明显，可快速交货。',
    attachments: [],
    createdAt: ts(-17, 10, 0),
    updatedAt: ts(-14, 15, 0),
  },
];

File: backend/app/routers/quotations.py
"""报价单路由：list / get / create / saveDraft / submit

注：GET /inquiries/{inquiryId}/quotations 在 inquiries 路由中实现（前缀归属）
"""
from typing import Optional

from fastapi import APIRouter, Body, Depends, HTTPException, Query, status
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session, joinedload

from ..database import get_db
from ..idempotency import get_result, store_result
from ..models import User, Quotation, QuotationItem, Inquiry, InquiryLog
from ..schemas import (
    QuotationSchema, QuotationCreate, QuotationDraft, SuccessResult,
    PaginatedQuotationsSchema,
)
from ..auth import get_current_user
from ..policy import require_inquiry_access
from ..serializers import quotation_to_schema, gen_id, now_str
from ..events import publish

router = APIRouter(prefix="/quotations", tags=["quotations"])

QUOTATION_SUBMITTED = "SUBMITTED"
LOG_TYPE_SUBMIT_QUOTATION = "SUBMIT_QUOTATION"


def _get_quotation_with_access(db: Session, user: User, quotation_id: str) -> Quotation:
    """按 ID 加载报价单并做组织级访问校验（P1：报价单属询价单，复用 require_inquiry_access）。

    报价单本身无 organization 字段，其归属由父询价单决定；因此加载报价后校验父询价单，
    用户无权访问该询价单时返回 403，防止跨组织 IDOR。
    """
    q = db.query(Quotation).filter(Quotation.id == quotation_id).first()
    if q is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="报价单不存在")
    inq = db.query(Inquiry).filter(Inquiry.id == q.inquiry_id).first()
    if inq is None:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="无权访问该报价单",
        )
    require_inquiry_access(user, inq)
    return q


def _build_quotation_from_data(data: dict) -> Quotation:
    """从请求 dict 构造 Quotation ORM（含 items）"""
    q = Quotation(
        id=data.get("id") or gen_id("q"),
        inquiry_id=data["inquiryId"],
        supplier_id=data["supplierId"],
        supplier_name=data.get("supplierName", ""),
        status=data.get("status", "DRAFT"),
        submitted_at=data.get("submittedAt"),
        total_amount=data.get("totalAmount", 0),
        remark=data.get("remark"),
        created_at=data.get("createdAt") or now_str(),
        updated_at=data.get("updatedAt") or now_str(),
    )
    for item_data in data.get("items", []) or []:
        q.items.append(QuotationItem(
            id=item_data.get("id") or gen_id("qitem"),
            quotation_id=q.id,
            inquiry_item_id=item_data["inquiryItemId"],
            unit_price=item_data.get("unitPrice", 0),
            tax_rate=item_data.get("taxRate", 0),
            tax_included_total=item_data.get("taxIncludedTotal", 0),
            moq=item_data.get("moq"),
            delivery_days=item_data.get("deliveryDays", 0),
            delivery_date=item_data.get("deliveryDate"),
            brand=item_data.get("brand"),
            warranty_months=item_data.get("warrantyMonths"),
            payment_terms=item_data.get("paymentTerms"),
            valid_until=item_data.get("validUntil"),
            tech_deviation=item_data.get("techDeviation"),
            commercial_deviation=item_data.get("commercialDeviation"),
            remark=item_data.get("remark"),
        ))
    return q


@router.get("")
def list_quotations(
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
    page: Optional[int] = Query(default=None, ge=1),
    pageSize: Optional[int] = Query(default=None, ge=1, le=200),
):
    """报价列表（P2 Task 22：服务端分页 + 稳定排序 + joinedload 消除 items N+1）

    - 向后兼容：不传分页参数时返回全量数组（保持既有调用不变）。
    - 传入 page/pageSize 时返回分页结构 {items, total, page, pageSize}。
    """
    # Task 7 / Task 22：稳定排序（created_at,id 双键）+ joinedload 消除 items N+1
    query = (
        db.query(Quotation)
        .options(joinedload(Quotation.items))
        .order_by(Quotation.created_at, Quotation.id)
    )
    if page is None and pageSize is None:
        rows = query.all()
        return [quotation_to_schema(q, db) for q in rows]

    _page = page if page is not None else 1
    _size = pageSize if pageSize is not None else 20
    total = db.query(Quotation).count()
    rows = query.offset((_page - 1) * _size).limit(_size).all()
    return PaginatedQuotationsSchema(
        items=[quotation_to_schema(q, db) for q in rows],
        total=total,
        page=_page,
        pageSize=_size,
    )


@router.get("/{quotation_id}", response_model=QuotationSchema)
def get_quotation(
    quotation_id: str,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    q = _get_quotation_with_access(db, user, quotation_id)
    return quotation_to_schema(q, db)


@router.post("", response_model=QuotationSchema)
def create_quotation(
    body: QuotationCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    data = body.model_dump()
    # 组织级校验：报价必须归属当前用户可访问的询价单
    inq = db.query(Inquiry).filter(Inquiry.id == data["inquiryId"]).first()
    if inq is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="询价单不存在")
    require_inquiry_access(user, inq)
    q = _build_quotation_from_data(data)
    db.add(q)
    try:
        db.commit()
    except IntegrityError:
        # Task 6：uq_quotations_inquiry_id_supplier_id 唯一约束兜底，重复提交返回 409
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "error_type": "duplicate_quotation",
                "message": "该供应商已存在此询价单的报价，请勿重复创建",
                "inquiryId": data["inquiryId"],
                "supplierId": data.get("supplierId"),
            },
        )
    db.refresh(q)
    return quotation_to_schema(q, db)


@router.put("/{quotation_id}/draft", response_model=QuotationSchema)
def save_draft(
    quotation_id: str,
    body: QuotationDraft,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    q = _get_quotation_with_access(db, user, quotation_id)
    data = body.model_dump(exclude_unset=True)
    # 更新标量字段
    scalar_map = {
        "supplierId": "supplier_id", "supplierName": "supplier_name",
        "status": "status", "submittedAt": "submitted_at",
        "totalAmount": "total_amount", "remark": "remark",
    }
    for camel, snake in scalar_map.items():
        if camel in data:
            setattr(q, snake, data[camel])
    # 更新 items（整体替换）
    if "items" in data:
        for old_item in q.items:
            db.delete(old_item)
        q.items = []
        for item_data in data["items"] or []:
            q.items.append(QuotationItem(
                id=item_data.get("id") or gen_id("qitem"),
                quotation_id=q.id,
                inquiry_item_id=item_data["inquiryItemId"],
                unit_price=item_data.get("unitPrice", 0),
                tax_rate=item_data.get("taxRate", 0),
                tax_included_total=item_data.get("taxIncludedTotal", 0),
                moq=item_data.get("moq"),
                delivery_days=item_data.get("deliveryDays", 0),
                delivery_date=item_data.get("deliveryDate"),
                brand=item_data.get("brand"),
                warranty_months=item_data.get("warrantyMonths"),
                payment_terms=item_data.get("paymentTerms"),
                valid_until=item_data.get("validUntil"),
                tech_deviation=item_data.get("techDeviation"),
                commercial_deviation=item_data.get("commercialDeviation"),
                remark=item_data.get("remark"),
            ))
    q.updated_at = now_str()
    db.commit()
    db.refresh(q)
    return quotation_to_schema(q, db)


@router.post("/{quotation_id}/submit", response_model=QuotationSchema)
def submit_quotation(
    quotation_id: str,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
    body: dict = Body(default=None),
):
    """提交报价：status→SUBMITTED, submittedAt=now, 追加 SUBMIT_QUOTATION 日志到对应 inquiry

    Task 6 幂等：请求体可携带 idempotencyKey，已处理过则直接返回缓存结果，避免重复日志/重复事件。
    """
    idem_key = body.get("idempotencyKey") if body else None
    if idem_key:
        cached = get_result(idem_key, "quotations.submit")
        if cached is not None:
            return cached

    q = _get_quotation_with_access(db, user, quotation_id)
    ts = now_str()
    q.status = QUOTATION_SUBMITTED
    q.submitted_at = ts
    q.updated_at = ts
    # 追加日志到父询价单
    inquiry = db.query(Inquiry).filter(Inquiry.id == q.inquiry_id).first()
    if inquiry is not None:
        log = InquiryLog(
            id=gen_id(f"log-{inquiry.id}"),
            inquiry_id=inquiry.id,
            time=ts,
            operator=q.supplier_name,
            operator_role="供应商",
            type=LOG_TYPE_SUBMIT_QUOTATION,
            content="提交报价",
        )
        db.add(log)
        inquiry.updated_at = ts
    db.commit()
    db.refresh(q)
    # P2-12 Task 17：广播报价提交事件，SSE 订阅端据此刷新未读/详情/比价/通知中心
    publish("quotation_submitted", {"quotationId": q.id, "inquiryId": q.inquiry_id, "supplierId": q.supplier_id})
    result = quotation_to_schema(q, db)
    if idem_key:
        store_result(idem_key, "quotations.submit", result)
    return result
